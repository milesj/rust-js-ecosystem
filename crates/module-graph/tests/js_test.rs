mod utils;

use starbase_sandbox::{assert_snapshot, create_sandbox};
use utils::*;

mod mjs {
    use super::*;

    #[test]
    fn export_star() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-star.mjs"
        ));
    }

    #[test]
    fn export_named() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-named.mjs"
        ));
    }

    #[test]
    fn export_default_function() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-def-func.mjs"
        ));
        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-def-anon-func.mjs"
        ));
    }

    #[test]
    fn export_default_class() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-def-class.mjs"
        ));
        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-def-anon-class.mjs"
        ));
    }

    #[test]
    fn export_default_reference() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/export-def-ref.mjs"
        ));
    }

    #[test]
    fn dynamic_import_top_level_await() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/dyn-import-tla.mjs"
        ));
    }

    #[test]
    fn dynamic_import_scope_depths() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/dyn-import-scopes.mjs"
        ));
    }

    #[test]
    fn dynamic_import_destructure_patterns() {
        let sandbox = create_sandbox("js");

        assert_snapshot!(generate_graph_for_file(
            sandbox.path(),
            "mjs/dyn-import-patterns.mjs"
        ));
    }
}
