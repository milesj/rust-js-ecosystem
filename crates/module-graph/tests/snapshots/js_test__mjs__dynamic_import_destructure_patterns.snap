---
source: crates/module-graph/tests/js_test.rs
expression: "generate_graph_for_file(sandbox.path(), \"mjs/dyn-import-patterns.mjs\")"
---
{
    2: Module {
        exports: [],
        fragment: None,
        id: 2,
        imports: [],
        package_name: None,
        path: "/private/root/mjs/shared-def-object.mjs",
        query: None,
        source: JavaScriptModule {
            package_type: Unknown,
            source: "export default {\n  number: 123,\n  string: \"abc\",\n  object: {\n    one: {\n      two: {\n        value: true,\n      },\n    },\n    value: true,\n  },\n};\n",
            source_type: SourceType {
                language: JavaScript,
                module_kind: Module,
                variant: Jsx,
                always_strict: false,
            },
            stats: JavaScriptStats {
                dynamic_import_count: 0,
                export_statements: 1,
                exports_default: false,
                import_statements: 0,
                other_statements: 0,
                require_count: 0,
            },
        },
    },
    3: Module {
        exports: [
            Export {
                kind: Modern,
                module_id: None,
                source: None,
                span: Some(
                    Span {
                        start: 0,
                        end: 26,
                    },
                ),
                symbols: [
                    ExportedSymbol {
                        kind: Value,
                        symbol_id: None,
                        name: "number",
                    },
                ],
                type_only: false,
            },
            Export {
                kind: Modern,
                module_id: None,
                source: None,
                span: Some(
                    Span {
                        start: 27,
                        end: 55,
                    },
                ),
                symbols: [
                    ExportedSymbol {
                        kind: Value,
                        symbol_id: None,
                        name: "string",
                    },
                ],
                type_only: false,
            },
        ],
        fragment: None,
        id: 3,
        imports: [],
        package_name: None,
        path: "/private/root/mjs/shared.mjs",
        query: None,
        source: JavaScriptModule {
            package_type: Unknown,
            source: "export const number = 123;\nexport const string = \"abc\";\n",
            source_type: SourceType {
                language: JavaScript,
                module_kind: Module,
                variant: Jsx,
                always_strict: false,
            },
            stats: JavaScriptStats {
                dynamic_import_count: 0,
                export_statements: 2,
                exports_default: false,
                import_statements: 0,
                other_statements: 0,
                require_count: 0,
            },
        },
    },
    1: Module {
        exports: [],
        fragment: None,
        id: 1,
        imports: [
            Import {
                kind: AsyncDynamic,
                module_id: 2,
                source_request: "./shared-def-object.mjs",
                span: Span {
                    start: 42,
                    end: 75,
                },
                symbols: [
                    ImportedSymbol {
                        kind: Default,
                        source_name: Some(
                            "default",
                        ),
                        symbol_id: None,
                        name: "def",
                    },
                ],
                type_only: false,
            },
            Import {
                kind: AsyncDynamic,
                module_id: 3,
                source_request: "./shared.mjs",
                span: Span {
                    start: 118,
                    end: 140,
                },
                symbols: [
                    ImportedSymbol {
                        kind: Value,
                        source_name: None,
                        symbol_id: None,
                        name: "number",
                    },
                ],
                type_only: false,
            },
            Import {
                kind: AsyncDynamic,
                module_id: 3,
                source_request: "./shared.mjs",
                span: Span {
                    start: 202,
                    end: 224,
                },
                symbols: [
                    ImportedSymbol {
                        kind: Value,
                        source_name: None,
                        symbol_id: None,
                        name: "string",
                    },
                    ImportedSymbol {
                        kind: Namespace,
                        source_name: None,
                        symbol_id: None,
                        name: "rest",
                    },
                ],
                type_only: false,
            },
            Import {
                kind: AsyncDynamic,
                module_id: 2,
                source_request: "./shared-def-object.mjs",
                span: Span {
                    start: 359,
                    end: 392,
                },
                symbols: [
                    ImportedSymbol {
                        kind: Default,
                        source_name: None,
                        symbol_id: None,
                        name: "default",
                    },
                ],
                type_only: false,
            },
            Import {
                kind: AsyncDynamic,
                module_id: 3,
                source_request: "./shared.mjs",
                span: Span {
                    start: 459,
                    end: 481,
                },
                symbols: [
                    ImportedSymbol {
                        kind: Value,
                        source_name: None,
                        symbol_id: None,
                        name: "unknown",
                    },
                ],
                type_only: false,
            },
        ],
        package_name: None,
        path: "/root/mjs/dyn-import-patterns.mjs",
        query: None,
        source: JavaScriptModule {
            package_type: Unknown,
            source: "// Default\nconst { default: def } = await import(\"./shared-def-object.mjs\");\n\n// Destructure\nconst { number } = await import(\"./shared.mjs\");\n\n// Destructure with rest\nconst { string, ...rest } = await import(\"./shared.mjs\");\n\n// Destructure nested object\nconst {\n  default: {\n    object: {\n      one: {\n        two: { value },\n      },\n    },\n  },\n} = await import(\"./shared-def-object.mjs\");\n\n// Assignment, unknown export\nconst { unknown = \"abc\" } = await import(\"./shared.mjs\");\n",
            source_type: SourceType {
                language: JavaScript,
                module_kind: Module,
                variant: Jsx,
                always_strict: false,
            },
            stats: JavaScriptStats {
                dynamic_import_count: 5,
                export_statements: 0,
                exports_default: false,
                import_statements: 0,
                other_statements: 5,
                require_count: 0,
            },
        },
    },
}
